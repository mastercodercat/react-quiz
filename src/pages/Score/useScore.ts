import { useMemo } from "react";
import { useHistory } from "react-router-dom";

import { useAppSelector, useAppDispatch } from "store/hooks";
import { RootState } from "store/store";
import { fetchQuestions } from "store/quiz";

const useScore = () => {
  const dispatch = useAppDispatch();
  const { questions, answers } = useAppSelector(
    (state: RootState) => state.quiz
  );
  const history = useHistory();

  const correctCount = useMemo(() => {
    let count = 0;
    for (let i = 0; i < questions.length; i++) {
      if (questions[i].correct_answer === answers[i]) {
        count++;
      }
    }
    return count;
  }, [questions, answers]);

  const handleGoHome = () => {
    history.push("/");
  };

  const handlePlayAgain = async () => {
    await dispatch(fetchQuestions());
    history.push("/quiz/1");
  };

  return {
    correctCount,
    questions,
    answers,
    handleGoHome,
    handlePlayAgain,
  };
};

export default useScore;
