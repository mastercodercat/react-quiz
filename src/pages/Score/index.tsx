import { Button } from "react-bootstrap";

import Layout from "components/Layout";
import ResultItem from "components/ResultItem";

import useScore from "./useScore";

const Score = () => {
  const { correctCount, questions, answers, handleGoHome, handlePlayAgain } =
    useScore();

  return (
    <Layout title={`You scored ${correctCount}/${questions.length}`}>
      {questions.map((question, index) => (
        <ResultItem key={index} question={question} answer={answers[index]} />
      ))}
      <div className="mt-5 d-flex justify-content-between">
        <Button variant="success" onClick={handleGoHome}>
          Go Home
        </Button>
        <Button variant="primary" onClick={handlePlayAgain}>
          Play Again
        </Button>
      </div>
    </Layout>
  );
};

export default Score;
