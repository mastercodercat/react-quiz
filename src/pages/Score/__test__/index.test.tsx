import {
  screen,
  act,
  waitFor,
  fireEvent,
  RenderResult,
} from "@testing-library/react";

import renderTest from "test/helpers/renderTest";
import Score from "..";

let wrapper: RenderResult;

describe("Score", () => {
  beforeEach(() => {
    jest.useFakeTimers();
    wrapper = renderTest(<Score />, { routes: ["/score"] }, "/score");
  });

  afterEach(() => {
    act(() => {
      jest.runOnlyPendingTimers();
    });
    jest.useRealTimers();
  });

  it("should test snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should display score board", () => {
    expect(screen.queryByText("You scored 0/10")).toBeInTheDocument();
    expect(screen.queryByText("Go Home")).toBeInTheDocument();
    expect(screen.queryByText("Play Again")).toBeInTheDocument();
  });

  it("should go home after clicking button", async () => {
    fireEvent.click(screen.queryByText("Go Home") as Element);

    await waitFor(() => {
      expect(
        screen.queryByText("Welcome to the Trivia Challenge!")
      ).toBeInTheDocument();
    });
  });
});
