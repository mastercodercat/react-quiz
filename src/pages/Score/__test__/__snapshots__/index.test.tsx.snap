// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Score should test snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="my-5 mx-auto p-5 bg-secondary bg-opacity-25 col-lg-6 col-md-8 col-sm-12"
        >
          <div
            class="mb-5 row"
          >
            <div
              class="text-center col-sm-12"
            >
              <h1>
                You scored 0/10
              </h1>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-sm-12"
            >
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                This is the correct spelling of Supercalifragilisticexpialidocious.
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                The Kingdom of Prussia briefly held land in Estonia.
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                "Number 16 Bus Shelter" was a child's name that was approved by the New Zealand government.
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                The Battle of Trafalgar took place on October 23rd, 1805
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                Lady Gaga's real name is Stefani Joanne Angelina Germanotta.
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                During the Winter War, the amount of Soviet Union soliders that died or went missing was five times more than Finland's.
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                Janus was the Roman god of doorways and passageways.
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                The video game "Fuel" has an open world that is 5,560 square miles?
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                Only one country in the world starts with the letter Q.
              </div>
              <div
                class="text-white result-item bg-danger list-group-item"
              >
                The protagonist in "Humanity Has Declined" has no discernable name and is simply referred to as 'I' for most of the series.
              </div>
              <div
                class="mt-5 d-flex justify-content-between"
              >
                <button
                  class="btn btn-success"
                  type="button"
                >
                  Go Home
                </button>
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Play Again
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="my-5 mx-auto p-5 bg-secondary bg-opacity-25 col-lg-6 col-md-8 col-sm-12"
      >
        <div
          class="mb-5 row"
        >
          <div
            class="text-center col-sm-12"
          >
            <h1>
              You scored 0/10
            </h1>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-12"
          >
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              This is the correct spelling of Supercalifragilisticexpialidocious.
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              The Kingdom of Prussia briefly held land in Estonia.
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              "Number 16 Bus Shelter" was a child's name that was approved by the New Zealand government.
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              The Battle of Trafalgar took place on October 23rd, 1805
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              Lady Gaga's real name is Stefani Joanne Angelina Germanotta.
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              During the Winter War, the amount of Soviet Union soliders that died or went missing was five times more than Finland's.
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              Janus was the Roman god of doorways and passageways.
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              The video game "Fuel" has an open world that is 5,560 square miles?
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              Only one country in the world starts with the letter Q.
            </div>
            <div
              class="text-white result-item bg-danger list-group-item"
            >
              The protagonist in "Humanity Has Declined" has no discernable name and is simply referred to as 'I' for most of the series.
            </div>
            <div
              class="mt-5 d-flex justify-content-between"
            >
              <button
                class="btn btn-success"
                type="button"
              >
                Go Home
              </button>
              <button
                class="btn btn-primary"
                type="button"
              >
                Play Again
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
